programs/mango-v4/src/instructions/account_buyback_fees_with_mngo.rs

```diff
@@ -0,0 +1,151 @@
+    ctx: Context<AccountBuybackFeesWithMngo>,
+    max_buyback: u64,
+) -> Result<()> {
+    // Cannot buyback from yourself
+    require_keys_neq!(
+        ctx.accounts.account.key(),
+        ctx.accounts.dao_account.key(),
+        MangoError::SomeError
+    );
+
+    let mut account = ctx.accounts.account.load_full_mut()?;
+    // account constraint #1
+    require!(
+        account.fixed.is_owner_or_delegate(ctx.accounts.owner.key()),
+        MangoError::SomeError
+    );
+
+    let mut dao_account = ctx.accounts.dao_account.load_full_mut()?;
+
+    let group = ctx.accounts.group.load()?;
+
+    let mut mngo_bank = ctx.accounts.mngo_bank.load_mut()?;
+    let mut fees_bank = ctx.accounts.fees_bank.load_mut()?;
+
+    let bonus_rate = I80F48::from_num(group.fees_mngo_bonus_factor);
+    let now_ts = Clock::get()?.unix_timestamp.try_into().unwrap();
+
+    // quick return if nothing to buyback
+    let mut max_buyback =
+        I80F48::from_num::<u64>(max_buyback.min(account.fixed.discount_buyback_fees_accrued));
+    if max_buyback == I80F48::ZERO {
+        msg!(
+            "nothing to buyback, (discount_buyback_fees_accrued {})",
+            account.fixed.discount_buyback_fees_accrued
+        );
+        return Ok(());
+    }
+
+    // if mngo token position has borrows, skip buyback
+    let (account_mngo_token_position, account_mngo_raw_token_index, _) =
+        account.ensure_token_position(mngo_bank.token_index)?;
+    let account_mngo_native = account_mngo_token_position.native(&mngo_bank);
+    if account_mngo_native.is_negative() {
+        msg!(
+            "account mngo token position ({} native mngo) has borrows, nothing will be bought back",
+            account_mngo_native
+        );
+        return Ok(());
+    }
+
+    // compute max mngo to swap for fees
+    let mngo_oracle_price = mngo_bank.oracle_price(
+        &AccountInfoRef::borrow(&ctx.accounts.mngo_oracle.as_ref())?,
+        Some(Clock::get()?.slot),
+    )?;
+    let mngo_buyback_price = cm!(mngo_oracle_price * bonus_rate);
+    // mngo is exchanged at a discount
+    let mut max_buyback_mngo = cm!(max_buyback / mngo_buyback_price);
+    // buyback is restricted to accounts token position
+    max_buyback_mngo = max_buyback_mngo.min(account_mngo_native);
+    max_buyback = cm!(max_buyback_mngo * mngo_buyback_price);
+
+    // move mngo from user to dao
+    let (dao_mngo_token_position, dao_mngo_raw_token_index, _) =
+        dao_account.ensure_token_position(mngo_bank.token_index)?;
+    let in_use = mngo_bank.withdraw_without_fee(
+        account_mngo_token_position,
+        max_buyback_mngo,
+        now_ts,
+        mngo_oracle_price,
+    )?;
+    if !in_use {
+        account.deactivate_token_position_and_log(
+            account_mngo_raw_token_index,
+            ctx.accounts.account.key(),
+        );
+    }
+    let in_use = mngo_bank.deposit(dao_mngo_token_position, max_buyback_mngo, now_ts)?;
+    if !in_use {
```

Comment:
this `in_use` check we could skip by requiring that the dao_account's mngo token position is >= 0 beforehand